File: Java/RotateList.java
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) {
        this.val = val;
    }
    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

public class RotateList {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null) {
            return head;
        }

        int length = 1;
        ListNode dummy = head;
        while (dummy.next != null) {
            dummy = dummy.next;
            length++;
        }

        int pos = k % length;
        if (pos == 0) {
            return head;
        }

        ListNode current = head;
        for (int i = 0; i < length - pos - 1; i++) {
            current = current.next;
        }

        ListNode new_head = current.next;
        current.next = null;
        dummy.next = head;

        return new_head;
    }

    public static ListNode createList(int[] arr) {
        if (arr.length == 0) return null;
        ListNode head = new ListNode(arr[0]);
        ListNode temp = head;
        for (int i = 1; i < arr.length; i++) {
            temp.next = new ListNode(arr[i]);
            temp = temp.next;
        }
        return head;
    }

    public static void printList(ListNode head) {
        ListNode temp = head;
        while (temp != null) {
            System.out.print(temp.val + " -> ");
            temp = temp.next;
        }
        System.out.println("null");
    }

    public static void main(String[] args) {
        int[] arr = {1, 2, 3, 4, 5};
        int k = 2;

        ListNode head = createList(arr);

        System.out.println("Before rotation:");
        printList(head);

        RotateList sol = new RotateList();
        head = sol.rotateRight(head, k);

        System.out.println("After rotation:");
        printList(head);
    }
}

File: LeetCode/Python/RotateList.py
from typing import Optional

class ListNode:
    def __init__(self,val=0,next=None):
        self.val=val
        self.next=next

def rotateRight(head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head:
            return None
        length=1
        dummy=head
        while dummy.next:
            dummy=dummy.next
            length+=1
        pos=k%length

        if pos==0:
            return head
        current=head
        for _ in range(length-pos-1):
            current=current.next

        new_head=current.next
        current.next=None
        dummy.next=head

        return new_head
    

def print_ll(head):
    temp=head
    while temp:
        print(temp.val,end=' -> ')
        temp=temp.next
    print(None)
    
def list_to_ll(ls):
    head=ListNode(ls[0],None)
    temp=head
    for i in range(1,len(ls)):
        temp.next=ListNode(ls[i])
        temp=temp.next
    return head

ls=[1,2,3,4,5]
k=2
ls=list_to_ll(ls)
print('Before rotation :')
print_ll(ls)
print('\nk :',k)
ans=rotateRight(ls,k)
print('\nAfter rotation')
print_ll(ans)


# Input: head = [1,2,3,4,5], k = 2
# Output: [4,5,1,2,3]

# Input: head = [0,1,2], k = 4
# Output: [2,0,1]
 